python/graphstorm/dataloading/sampler.py

```diff
@@ -389,3 +392,247 @@ def sample_blocks(self, g, seed_nodes, exclude_eids=None):
+        self.reset = True
+        self.is_train = is_train
+
+    def __iter__(self):
+        if self.reset:
+            self._index = -1
+        self.reset = True
+        return self
+
+    def __next__(self):
+        """ Get index for next file"""
+        self._index += 1
+
+        if self._index >= self.num_local_files:
+            if self.is_train:
+                self._index = 0
+            else:
+                # non-infinite sampler if it's for evaluation.
+                self._index = 0
+                return None
+
+        return self.file_indices[self._index]
+
+class RandomShuffleFileSampler():
+    r"""Random File Sampler. Has files reshuffled for sampling.
+
+    Parameters:
+    ----------
+    file_indices : list of int
+        File indices for a local trainer
+    """
+
+    def __init__(self, file_indices=None):
+        self.file_indices = file_indices
+        self.num_local_files = len(self.file_indices)
+        self._indices = list(range(self.num_local_files))
+        self._index = -1
+        self.reset = True
+
+    def __iter__(self):
+        if self.reset:
+            self._index = -1
+            # shuffle the file indices
+            random.shuffle(self._indices)
+        self.reset = True
+        return self
+
+    def __next__(self):
+        """ Get index for next file"""
+        self._index += 1
+
+        if self._index >= self.num_local_files:
+            # make it infinite sampler
+            self._index = 0
+
+        return self.file_indices[self._indices[self._index]]
+
+class DistributedFileSampler(_FileSampler):
+    r"""Distributed File Sampler. Samples file from the data path.
+    Each rank only has access to a partition of all the data shards.
+
+    Parameters:
+    ---------
+    dataset_path : str
+        path to the data files.
+        Mutually exclusive with :attr:`files`.
+    shuffle : bool
+        Set to ``True`` to have the files reshuffled
+        at every epoch. Shuffling is performed on the files of the local partition.
+    infinite : bool
+        Set to ``True`` to have the files sampled infinitely.
+    local_rank : int
+        Local rank ID.
+    world_size : int
+        Number of all trainers.
+    is_train : bool
+        Set to ``True`` if it's training set.
+    """
+
+    def __init__(
+        self, 
+        dataset_path=None, 
+        shuffle=False, 
+        infinite=False, 
+        local_rank=-1, 
+        world_size=None, 
+        is_train=True,
+    ):
+        super().__init__(dataset_path)
+
+        # Initialize distributed ranks
+        if local_rank == -1:
+            self.global_rank = 0
+            self.dp_size = 1
+            self.dp_rank = self.global_rank
+        else:
+            self.global_rank = gs.get_rank()
+            self.dp_rank = local_rank
+            self.dp_size = world_size
+
+        if self.dp_size > self.num_files:
+            self.remainder = self.dp_size % self.num_files
+            self.part_start = 0
```

Comment:
Can you explain why start is 0 and end is num_files?
Does each trainer use the entire training set?
Please add comments here.